dagjose/multicodec.go

```diff
@@ -1,11 +1,127 @@
 package dagjose
 
+//go:generate go run ./gen .
+//go:generate go fmt ./
+
 import (
-	dagcbor "github.com/ipld/go-ipld-prime/codec/dagcbor"
-	multicodec "github.com/ipld/go-ipld-prime/multicodec"
+	"errors"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/codec"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/multicodec"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/multiformats/go-multibase"
 )
 
 func init() {
-	multicodec.RegisterDecoder(0x85, dagcbor.Decode)
-	multicodec.RegisterEncoder(0x85, dagcbor.Encode)
+	multicodec.RegisterDecoder(0x85, Decode)
+	multicodec.RegisterEncoder(0x85, Encode)
+}
+
+// Decode deserializes data from the given io.Reader and feeds it into the given datamodel.NodeAssembler. Decode fits
+// the codec.Decoder function interface.
+func Decode(na datamodel.NodeAssembler, r io.Reader) error {
+	// If the passed `NodeAssembler` is not of type `_DecodedJOSE__ReprBuilder`, create and use a
+	// `_DecodedJOSE__ReprBuilder`.
+	joseBuilder, alreadyJose := na.(*_DecodedJOSE__ReprBuilder)
+	if !alreadyJose {
+		joseBuilder = Type.DecodedJOSE__Repr.NewBuilder().(*_DecodedJOSE__ReprBuilder)
+	}
```

Comment:
This successfully detects and shortcuts if someone hands in the builder form...

But not if they hand in the assembler, for whatever reason.

I had to double-check myself quick on how this works, and so came out with a gist: https://play.golang.org/p/zdvRuU3WY1g .  Moral of the gist is: you can certainly make code that works on either type, but unfortunately when doing the assertions to see what was handed _in_, you might want to do two of them, so either builder or assembler form will both hit the fastpath.

(My apologies for this being a little confusing.  A lot of this area was designed with "maximum fast" in mind, and the ease-of-use is correspondingly... well.  I wish I could think of a way to make this simpler!)

Reply:
Ahh, gotcha, will update, ty.

Reply:
Hmm, looks like the coercion [doesn't work](https://play.golang.org/p/osjyNvWT2wv) with the way the builder/assembler are generated 😞
```
type _DecodedJOSE__ReprBuilder struct {
	_DecodedJOSE__ReprAssembler
}
type _DecodedJOSE__ReprAssembler struct {
	thing int // n.b. same name "thing", same int type
}
```
```
./prog.go:37:41: cannot convert assembler (type *_DecodedJOSE__ReprAssembler) to type *_DecodedJOSE__ReprBuilder
```
versus the way it's in the example you sent:
```
type _DecodedJOSE__ReprBuilder struct {
	thing int
}
type _DecodedJOSE__ReprAssembler struct {
	thing int // n.b. same name "thing", same int type
}
```
Or is there some other way to do this that I'm missing?

Reply:
It works like this but I'm not sure if it like it:
```
builder := &_DecodedJOSE__ReprBuilder{*assembler}
_ = builder
```

Reply:
But, you know what? It might not be necessary to coerce the assembler into a builder.

The fastpath would be to pass either a JOSE builder or a JOSE assembler on to CBOR decode.

Could I do something like this?
```
	copyRequired := false
	jwsBuilder, castOk := na.(*_DecodedJWS__ReprBuilder)
	if !castOk {
		// This could still be `_DecodedJWS__ReprAssembler`, so check for that.
		_, castOk := na.(*_DecodedJWS__ReprAssembler)
		if !castOk {
			// No fastpath possible, just create a new `_DecodedJWE__ReprBuilder`, use it, then copy the built node into
			// the assembler the caller passed in.
			jwsBuilder = Type.DecodedJWS__Repr.NewBuilder().(*_DecodedJWS__ReprBuilder)
			copyRequired = true
		}
	}

	if err := dagcbor.Decode(jwsBuilder, r); err != nil {
		return err
	}
```